### ISSUE TITLE
Dashboards for dimmers

### ISSUE THREAD

**Issue by cprussin** [purplg/hass#29]
I can't seem to find any appropriate way of using dimmers with the dashboard.  I could possibly add a bunch of widgets, one for toggle, one for decreasing brightness, and one for increasing brightness -- but the latter two would require that I pass data with the service and I don't see a way to do that with the dashboard configuration options.  Am I missing something?  Is there some way to do this?

**Comment by purplg** [comment]
No, I don't think you're missing anything. I don't have any dimmers nor have I ever considered passing additional data to a service, but this is definitely a necessary feature. Thanks for reporting it!

Unfortunately, I'm not too happy with the implementation I have mocked up so far. Making something cleaner would require some breaking changes.

Try it out in the dev branch. I used a counter helper to test on. Here's an example where I set it's maximum:

```emacs-lisp
("counter.hass_counter"
 :label "Counter"
 :service "counter.configure"
 :data ,(json-encode '(("entity_id" . "counter.hass_counter")
                       ("maximum" . 5))))
```

Basically, I added the `:data` tag for the widget, which internally changes which service function it calls, which I definitely feel like can cause confusion to users.

Let me know if this is the functionality you were looking for and I'll work on it some more.

**Comment by cprussin** [comment]
@purplg Yeah, that looks exactly like what I had in mind.  I'd be happy to contribute back but I won't have a chance to hack on it any time super soon.

**Comment by cprussin** [comment]
Although now that I think about it, reusing a push-button widget is probably wrong.  Is it possible to use custom created widgets on the dashboard?  That's probably the right place to start here -- rather than making one widget that works for all cases, make each element use a definable widget, and provide widgets for the most common cases (for now, push buttons, and I can come back and contribute a slider when I have a chance).  Thoughts?

**Comment by purplg** [comment]
I had initially considered a new type of widget for this. Maybe a spinner of some sort with adjacent +/- buttons. I don't know how well a slider would work in Emacs, but I imagine if it's wide enough it could work. Or maybe SVGs could provide some level of resolution.

But yeah, I like the idea of new widget types.

In terms of the API, I suppose a new `:type` property could be added, width the default being the current `button`. Then different types could potentially support additional properties. For example, if a `spinner` widget type were created, `| - |  255  | + |`, I'd imagine the default `:service` would apply to the number readout, where it says 255, then there'd be the additional `:service-decrement` and `:service-increment` for the minus/plus buttons respectively.

Also, those should have derived defaults as well.

Anyway, those are my off-the-cuff thoughts.

**Comment by cprussin** [comment]
I was thinking something along these lines:

```elisp
(setq hass-dash-layout
      '((hass--group-widget "Group One"
                            ((hass--toggle-widget :entity "switch.bedroom_light")
                             (hass--toggle-widget :entity "switch.living_room_light")))
        (hass--group-widget "Group Two"
                            ((hass--group-widget "Nested Group"
                                                 ((hass--toggle-widget :entity "switch.kitchen_light")))))))
```

where:

- `hass-dash-layout` simply becomes a list of widgets
- `hass--group-widget` is a function that takes a name and a list of widgets and creates a new widget for that group
- `hass--toggle-widget` is basically the current `hass-dash--create-widget`

The downside is that it's a breaking change, the upside is that it enables the widget dashboard to become a tree of groups rather than a flat list of groups, and it enables people to write arbitrary functions to create custom widgets.

Thoughts?

**Comment by cprussin** [comment]
let me know if you like that proposal and I'll be happy to work on a PR @purplg ^

**Comment by purplg** [comment]
I do really like that. Much more descriptive and I think easier to work with. Yeah, go ahead. :)

**Comment by purplg** [comment]
@cprussin Since I recently got some new devices that support it, I've begun work on dimmer widgets. I call them "sliders" so the name applies to any kind of continuous widget. They should work for at least the `light` and `counter` domains and I still wanna give a percentage read out for things like lights, and a bunch of other little details.

You can check it out in dev if you'd like. You gotta bind `hass-dash-slider-decrease` and `hass-dash-slider-increase` in `hass-dash-mode-map` since I don't yet know sane defaults for that with vanilla Emacs keybindings (evil user). I just use `h` and `l` respectively.

**Comment by cprussin** [comment]
ahh sweet @purplg I'll check it out and will leave some feedback when I get a chance.  Sorry I never got to wrapping this up, I'd totally intended to but got side tracked by other projects and kept putting this on the backburner.

I'm actually an evil user as well so the `h` and `l` bindings are ideal for me too.  Maybe just bind em to `C-M-f` and `C-M-b` (forward and backward by sexp) for standard emacs?

**Comment by purplg** [comment]
No probs @cprussin I appreciate what you've done so far anyway :) Lemme know what you think or what you think could be different
